server/ocsp.go

```diff
@@ -0,0 +1,377 @@
+				t, nextRun,
+			)
+			continue
+		default:
+			s.Errorf("Received OCSP certificate status: %s, shutting down", ocspStatusString(n))
+			s.Shutdown()
+			return
+		}
+	}
+}
+
+// NewOCSPMonitor takes a TLS configuration then wraps it with the callbacks set for OCSP verification
+// along with a monitor that will periodically fetch OCSP staples.
+func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, error) {
+	opts := srv.getOpts()
+	oc := opts.OCSPConfig
+	tcOpts := opts.TLSConfigOpts
+	if oc.Mode == OCSPModeNever {
+		return tc, nil, nil
+	}
+	cert, err := tls.LoadX509KeyPair(tcOpts.CertFile, tcOpts.KeyFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if oc.Mode == OCSPModeAuto && !hasOCSPStatusRequest(cert.Leaf) {
+		// "status_request" flag not set. No need to do anything.
+		return tc, nil, nil
+	}
+
+	// TODO: Add OCSP 'responder_cert' option in case CA cert not available.
+	caCert := tcOpts.CaFile
+	issuer, err := getOCSPIssuer(caCert, cert.Certificate)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Callbacks below will be in charge of returning the certificate instead,
+	// so this has to be nil.
+	tc.Certificates = nil
+
+	mon := &OCSPMonitor{
+		srv: srv,
+		hc:  &http.Client{Timeout: 30 * time.Second},
+		// FIXME: Make reloadable aware, otherwise a change in certificates
+		//  could make the staples stale.
+		Leaf:   cert.Leaf,
+		Issuer: issuer,
+	}
+
+	// Get the certificate status from the memory, disk, then remote OCSP responder.
+	_, resp, err := mon.getStatus()
+	if err != nil {
+		return nil, nil, fmt.Errorf("bad OCSP status update: %s", err)
+	}
+	if err == nil && resp.Status != ocsp.Good && (oc.Mode == OCSPModeAlways || oc.Mode == OCSPModeMust) {
+		return nil, nil, fmt.Errorf("found existing OCSP certificate status: %s", ocspStatusString(resp.Status))
+	}
+
+	// GetCertificate returns a certificate that's presented to a
+	// client.
+	tc.GetCertificate = func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		raw, _, err := mon.getStatus()
+		if err != nil {
+			return nil, err
+		}
+
+		return &tls.Certificate{
+			OCSPStaple:                   raw,
+			Certificate:                  cert.Certificate,
+			PrivateKey:                   cert.PrivateKey,
+			SupportedSignatureAlgorithms: cert.SupportedSignatureAlgorithms,
+			SignedCertificateTimestamps:  cert.SignedCertificateTimestamps,
+			Leaf:                         cert.Leaf,
+		}, nil
+	}
+
+	// GetClientCertificate returns a certificate that's presented to a
+	// server.
+	tc.GetClientCertificate = func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return &cert, nil
+	}
+	return tc, mon, nil
+}
+
+func hasOCSPStatusRequest(cert *x509.Certificate) bool {
+	tlsFeatures := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24}
```

Comment:
Comment mentioning RFC 7633 as the reference and noting that this corresponds to `id-pe-tlsfeature` ?

Reply:
Cool, now documented: https://github.com/nats-io/nats-server/pull/2210/commits/ec5bfb9470aad3a9d8681bbc3d2bb316f2a7e2f9